### ISSUE TITLE
Provide a Way to Validate Templates

### ISSUE THREAD

**Issue by RehanSaeed** [dotnet/templating#367]
It's very easy to build and deploy a broken template. In particular, you might add complicated symbols that cause the project to fail to build when the user selects certain options. One way to help solve this is to validate the template. The validator would

1. Validate the `template.json` file.
2. Create the template with the default symbols, build it, maybe run it using the ASP.NET Core TestHost API to ensure it works.
3. Foreach symbol, provide a non-default value, then build it, maybe run it using the ASP.NET Core TestHost API.

I'm currently planning to do the above using xUnit tests. However, you might know a better way using the templating internals and Roslyn perhaps?

**Comment by mlorbetske** [comment]
@seancpeters had some thoughts on this on Friday, I'll let him speak to those when he gets the chance

**Comment by sayedihashimi** [comment]
I think having a `--validate` or `--inspect` would be valuable. But, I think it would be more valuable to have a JSON schema file that can be used. That would significantly decrease the need of validate/inspect and would simplify the authoring process of the config file. It would handle the static validation parts as well. I think the JSON schema would also likely be used by validate/inspect anyway so it's a stepping stone to that as well.

@RehanSaeed thoughts?

**Comment by RehanSaeed** [comment]
A JSON schema file is a great first step. --validate sounds great for a more in depth validation of the template, symbols and making sure the project builds.

**Comment by RehanSaeed** [comment]
I've written some [code](https://github.com/ASP-NET-Core-Boilerplate/Templates/tree/master/Tests/Boilerplate.Templates.Test) to run tests against a template that lets you do this:

```
[Fact]
public async Task Home_BuildsAndRuns_Returns200Ok()
{
    var projectDirectoryPath = TemplateAssert.GetProjectDirectoryPath(
        typeof(ApiTemplateTest).GetTypeInfo().Assembly,
        ProjectFileName);
    TemplateAssert.DotnetNewInstall(projectDirectoryPath).Wait();
    using (var project = await TemplateAssert.DotnetNew("bapi", "HomeTest"))
    {
        await TemplateAssert.DotnetRestore(project.DirectoryPath);
        await TemplateAssert.DotnetBuild(project.DirectoryPath);
        await TemplateAssert.DotnetPublish(project.DirectoryPath, "netcoreapp1.1");
        await TemplateAssert.NpmInstall(project.DirectoryPath);
        await TemplateAssert.Gulp(project.DirectoryPath, "build");
        await TemplateAssert.DotnetRun(
            project.DirectoryPath,
            async testServer =>
            {
                var response = await testServer.CreateClient().GetAsync("/");
                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
            });
    }
}
```

Right now it all works but with one issue remaining. `DotnetRun` dynamically loads the published assemblies and attempts to bootstrap the Startup class using the ASP.NET Core TestHost. Unfortunately, loading `Swashbuckle.AspNetCore.Swagger` causes a `TypeLoadException`. It seems to have trouble with third party libraries. I'd really appreciate any help as I'm at a loss.

**Comment by RehanSaeed** [comment]
If anyone is interested, I'll turn this into a NuGet package and get it released for public use but I'd like to get `DotnetRun` working first.

**Comment by mlorbetske** [comment]
If you'd like, it may be helpful to others to add this to the wiki - I've added a placeholder page for discussing ways to validate templates [here](https://github.com/dotnet/templating/wiki/Testing-your-templates). We'll add instructions for how to get in to the test battery that will run as part of the build of this repo (eventually, currently it's manually triggered) as well.

**Comment by RehanSaeed** [comment]
@mlorbetske My framework is still work in progress. I managed to get it working (including starting up an ASP.NET Core site using the `TestHost` API and executing some requests) but I had to restore, build and pubslish it as a standalone app, so it includes all the DLL's in one folder.

I currently have issue https://github.com/dotnet/corefx/issues/21982 open in the corefx repository to see if I can do the above using a normal non-standalone publish. Once I've sorted this issue, I'll release my code as a NuGet package for all to play use and profit from.
